src/codechicken/nei/BookmarkPanel.java

```diff
@@ -171,4 +173,68 @@ public int getWidth(GuiContainer gui) {
         return LayoutManager.getLeftSize(gui) - ( (gui.xSize + gui.width) / 2 + 3 ) - 16;
     }
 
+    @Override
+    public void mouseDragged(int mousex, int mousey, int button, long heldTime)
+    {
+
+        if (button == 0 && NEIClientUtils.shiftKey() && mouseDownSlot >= 0) {
+            ItemPanelSlot mouseOverSlot = getSlotMouseOver(mousex, mousey);
+
+            if (mouseOverSlot == null) {
+                return;
+            }
+
+            if (sortedStackIndex == -1) {
+                ItemStack stack = _items.get(mouseDownSlot);
+
+                if (stack != null && (mouseOverSlot == null || mouseOverSlot.slotIndex != mouseDownSlot || heldTime > 500)) {
+                    sortedStackIndex = mouseDownSlot;
+                }
+
+            } else if (mouseOverSlot != null && mouseOverSlot.slotIndex != sortedStackIndex) {
+
+                int maxStackIndex = Math.max(mouseOverSlot.slotIndex, sortedStackIndex);
+                int slotIndex = mouseOverSlot.slotIndex;
+                ItemStack cacheItem = _items.get(sortedStackIndex);
+                String recipeId = _recipes.get(sortedStackIndex);
+
+                if (sortedStackIndex == maxStackIndex) {
```

Comment:
Wouldn't a `_items.add(slotIndex, _items.remove(sortedStackIndex))` handle all of the shifting for you?

Reply:
thanks, I didn't know that this could be done